File: host/devflow-auth-host/auth_metadata.go
package main

import (
	"encoding/json"
	"time"
)

type SafeAuthMetadata struct {
	AccessExpiresAt      *string `json:"access_expires_at,omitempty"`
	RefreshExpiresAt     *string `json:"refresh_expires_at,omitempty"`
	HasRefreshCredential *bool   `json:"has_refresh_credential"`
	MetadataStatus       string  `json:"metadata_status"`
	RefreshExpirySource  string  `json:"refresh_expiry_source"`
}

// ExtractSafeMetadata securely extracts only non-sensitive expiration and presence metadata from credential secret.
// It never leaks tokens, passwords, or the raw secret payload.
func ExtractSafeMetadata(secretBytes []byte) SafeAuthMetadata {
	result := SafeAuthMetadata{
		MetadataStatus:      "unrecognized",
		RefreshExpirySource: "not_provided",
	}

	if len(secretBytes) == 0 {
		return result
	}

	var parsed map[string]interface{}
	if err := json.Unmarshal(secretBytes, &parsed); err != nil {
		// Not JSON or corrupted, return unrecognized safely
		return result
	}

	// 只有当存在有效 string 类型的 refresh_token 且非空时才设置 true；
	// 若明确为空字符串设置 false；若为 null、非 string 或未知类型，保留 nil，绝不武断推导为 false
	if val, ok := parsed["refresh_token"]; ok && val != nil {
		if str, isStr := val.(string); isStr {
			if len(str) > 0 {
				t := true
				result.HasRefreshCredential = &t
			} else {
				f := false
				result.HasRefreshCredential = &f
			}
		}
	}

	// Check access expiry
	var accessExpTime *time.Time
	if val, ok := parsed["expiry"]; ok && val != nil {
		if str, isStr := val.(string); isStr {
			if t, err := time.Parse(time.RFC3339, str); err == nil {
				accessExpTime = &t
			}
		}
	}
	if accessExpTime == nil {
		if val, ok := parsed["expires_at"]; ok && val != nil {
			if num, isNum := val.(float64); isNum && num > 0 {
				var t time.Time
				if num > 1e11 {
					t = time.UnixMilli(int64(num)).UTC()
				} else {
					t = time.Unix(int64(num), 0).UTC()
				}
				accessExpTime = &t
			} else if str, isStr := val.(string); isStr {
				if t, err := time.Parse(time.RFC3339, str); err == nil {
					accessExpTime = &t
				}
			}
		}
	}

	if accessExpTime != nil {
		s := accessExpTime.Format(time.RFC3339)
		result.AccessExpiresAt = &s
	}

	// Check refresh expiry if explicitly reported by provider
	var refreshExpTime *time.Time
	if val, ok := parsed["refresh_expires_at"]; ok && val != nil {
		if str, isStr := val.(string); isStr {
			if t, err := time.Parse(time.RFC3339, str); err == nil {
				refreshExpTime = &t
			}
		} else if num, isNum := val.(float64); isNum && num > 0 {
			var t time.Time
			if num > 1e11 {
				t = time.UnixMilli(int64(num)).UTC()
			} else {
				t = time.Unix(int64(num), 0).UTC()
			}
			refreshExpTime = &t
		}
	}

	if refreshExpTime != nil {
		s := refreshExpTime.Format(time.RFC3339)
		result.RefreshExpiresAt = &s
		result.RefreshExpirySource = "provider_reported"
	}

	// 校验 token 字段有效性：access_token 必须是非空 string，且不能是 null/空对象
	var hasValidAccessToken bool
	if tok, ok := parsed["access_token"]; ok && tok != nil {
		if str, isStr := tok.(string); isStr && len(str) > 0 {
			hasValidAccessToken = true
		}
	}
	var hasValidTokenType bool
	if tt, ok := parsed["token_type"]; ok && tt != nil {
		if str, isStr := tt.(string); isStr && len(str) > 0 {
			hasValidTokenType = true
		}
	}

	// 只有具备合法 token 字段且具备可验证的过期时间或明确的合法 refresh 存在性证明时，才认作 verified
	if (hasValidTokenType || hasValidAccessToken) && (result.AccessExpiresAt != nil || (result.HasRefreshCredential != nil && *result.HasRefreshCredential)) {
		result.MetadataStatus = "verified"
	}

	return result
}

File: packages/agy-accounts/src/auth-metadata.ts
import type { AgyAccountAuth } from "../../contracts/src/agy-account.js";

function tryExtractJwtClaims(token: string): { email?: string; subject?: string } {
  try {
    const parts = token.split(".");
    if (parts.length !== 3 || !parts[1]) return {};
    const payloadJson = Buffer.from(parts[1], "base64url").toString("utf8");
    const payload = JSON.parse(payloadJson) as Record<string, unknown>;
    if (!payload || typeof payload !== "object") return {};
    const res: { email?: string; subject?: string } = {};
    if (typeof payload.email === "string" && payload.email.includes("@")) {
      res.email = payload.email.trim();
    }
    if (typeof payload.sub === "string" && payload.sub.length > 0) {
      res.subject = payload.sub.trim();
    }
    return res;
  } catch {
    return {};
  }
}

/** Extract only the old host's explicit expiry/presence fields and identity claims, never raw token contents. */
export function extractSafeAuthMetadata(
  secret: Buffer,
): Partial<AgyAccountAuth> {
  const result: Partial<AgyAccountAuth> = {
    has_refresh_credential: null,
    metadata_status: "unrecognized",
    refresh_expiry_source: "not_provided",
  };
  let parsed: Record<string, unknown>;
  try {
    const value: unknown = JSON.parse(secret.toString("utf8"));
    if (!value || typeof value !== "object" || Array.isArray(value))
      return result;
    parsed = value as Record<string, unknown>;
  } catch {
    return result;
  }
  const date = (value: unknown): string | undefined => {
    if (typeof value === "string" && /^\d+(?:\.\d+)?$/.test(value))
      value = Number(value);
    if (typeof value !== "string" && typeof value !== "number")
      return undefined;
    if (typeof value === "number" && (!Number.isFinite(value) || value <= 0))
      return undefined;
    if (
      typeof value === "string" &&
      !/^\d{4}-\d\d-\d\dT.*(?:Z|[+-]\d\d:\d\d)$/.test(value)
    )
      return undefined;
    const ms =
      typeof value === "number"
        ? value > 1e11
          ? value
          : value * 1000
        : Date.parse(value);
    return Number.isFinite(ms) && !Number.isNaN(new Date(ms).getTime())
      ? new Date(ms).toISOString()
      : undefined;
  };
  if (typeof parsed.refresh_token === "string")
    result.has_refresh_credential = parsed.refresh_token.length > 0;
  const access =
    (typeof parsed.expiry === "string" ? date(parsed.expiry) : undefined) ??
    date(parsed.expires_at);
  const refresh = date(parsed.refresh_expires_at);
  if (access) result.access_expires_at = access;
  if (refresh) {
    result.refresh_expires_at = refresh;
    result.refresh_expiry_source = "provider_reported";
  }

  let email: string | undefined;
  let subject: string | undefined;
  if (typeof parsed.email === "string" && parsed.email.includes("@")) {
    email = parsed.email.trim();
  }
  if (typeof parsed.id_token === "string" && parsed.id_token.length > 0) {
    const claims = tryExtractJwtClaims(parsed.id_token);
    if (claims.email && !email) email = claims.email;
    if (claims.subject && !subject) subject = claims.subject;
  }
  if (!email && typeof parsed.access_token === "string" && parsed.access_token.length > 0) {
    const claims = tryExtractJwtClaims(parsed.access_token);
    if (claims.email) email = claims.email;
    if (claims.subject && !subject) subject = claims.subject;
  }
  if (email) result.email = email;
  if (subject) result.subject = subject;

  if (
    ((typeof parsed.access_token === "string" &&
      parsed.access_token.length > 0) ||
      (typeof parsed.token === "string" &&
        parsed.token.length > 0) ||
      (typeof parsed.id_token === "string" &&
        parsed.id_token.length > 0) ||
      (typeof parsed.token_type === "string" &&
        parsed.token_type.length > 0)) &&
    (access ||
      result.has_refresh_credential === true ||
      typeof parsed.id_token === "string" ||
      typeof parsed.token === "string")
  )
    result.metadata_status = "verified";
  return result;
}

